app/src/main/java/com/github/se/eventradar/viewmodel/qrCode/ScanFriendQrViewModel.kt

```diff
@@ -42,33 +65,100 @@ constructor(
     ScanQR
   }
 
-  private var myUID = ""
+  data class QrCodeScanFriendState(
+      val decodedResult: String = "",
+      val action: Action = Action.None,
+      val tabState: Tab = Tab.MyQR,
+      val username: String = "",
+      val qrCodeLink: String = "",
+      val isLoading: Boolean = true, // Indicates loading state
+  )
 
-  init {
-    viewModelScope.launch {
-      when (userRepository.getCurrentUserId()) {
-        is Resource.Success -> {
-          myUID = (userRepository.getCurrentUserId() as Resource.Success<String>).data
-        }
-        is Resource.Failure -> {
-          changeAction(Action.CantGetMyUID)
-        }
-      }
-    }
+  private var myUID: String = ""
 
-    qrCodeAnalyser.onDecoded = { decodedString ->
-      Log.d("QrCodeFriendViewModel", "Decoded QR Code: $decodedString")
-      val result = decodedString ?: "Failed to decode QR Code"
-      _uiState.value = _uiState.value.copy(decodedResult = result) // Update state flow
-      //      _decodedResult.value = result // Update state flow
-      if (result != "Failed to decode QR Code") {
-        updateFriendList(result) // Directly call updateFriendList
-      } else {
-        changeAction(Action.AnalyserError)
-      }
-    }
+  private val _uiState = MutableStateFlow(QrCodeScanFriendState())
+  val uiState: StateFlow<QrCodeScanFriendState> = _uiState
+
+  private val initialUiState: StateFlow<QrCodeScanFriendState> =
+      flow {
+            emit(QrCodeScanFriendState(isLoading = true))
+
+            when (val userIdResult = userRepository.getCurrentUserId()) {
+              is Resource.Success -> {
+                val getMyUID = userIdResult.data
+                //                emit(QrCodeScanFriendState(isLoading = true))
+                myUID = getMyUID
+              }
+              is Resource.Failure -> {
+                emit(QrCodeScanFriendState(isLoading = false, action = Action.CantGetMyUID))
+              }
+            }
+            val decodedResult =
+                callbackFlow {
+                      qrCodeAnalyser.onDecoded = { decodedString ->
+                        trySend(decodedString ?: "Failed to decode QR Code")
+                        close()
+                      }
+                      awaitClose { qrCodeAnalyser.onDecoded = null }
+                    }
+                    .first()
+
+            if (decodedResult == "Failed to decode QR Code") {
```

Comment:
I would try to change this part. It feels strange to introduce something that is not type safe in kotlin. Even just making it so that the default value is `null` would be better. If you *really* want to keep the string, you might want to create a constant for that

Reply:
okay, I changed it :)